src/ability/traits.rs

```diff
@@ -0,0 +1,79 @@
+use crate::{did::Did, nonce::Nonce, prove::TryProve};
+use libipld_cbor::DagCborCodec;
+use libipld_core::{
+    cid::{Cid, CidGeneric},
+    codec::Encode,
+    ipld::Ipld,
+    multihash::{Code::Sha2_256, MultihashDigest},
+    serde as ipld_serde,
+};
+use serde_derive::{Deserialize, Serialize};
+use std::{collections::BTreeMap, fmt::Debug};
+
+pub trait Command {
+    const COMMAND: &'static str;
+}
+
+// FIXME Delegable and make it proven?
+pub trait Delegatable: Sized {
+    type Builder: Debug + TryInto<Self> + From<Self>;
+}
+
+pub trait Resolvable: Sized {
+    type Awaiting: Debug + TryInto<Self> + From<Self>;
+}
+
+pub trait Runnable {
+    type Output: Debug;
+    fn task_id(self, subject: Did, nonce: Nonce) -> Cid;
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct DynJs {
+    pub cmd: String,
+    pub args: BTreeMap<String, Ipld>,
+}
+
+impl Delegatable for DynJs {
+    type Builder = Self;
+}
+
+impl Resolvable for DynJs {
+    type Awaiting = Self;
+}
+
+impl Runnable for DynJs {
+    type Output = Ipld;
+
+    fn task_id(self, subject: Did, nonce: Nonce) -> Cid {
```

Comment:
As per our discussion on needing this for what once was `instruction` within the data layout, shouldn't this be defaulted for non-DynJS implementations as well?

Reply:
It depends on how effectful it is. Sometimes you want the nonce, sometimes not.

Reply:
@expede yeah, I mean, it used to be that the nonce was defaulted I guess to `""`, so it was either there or had no effect. I'd say that then we'd have two functions then. It's crucial for the runtime, and the nonce is valuable *if* you want to run duplicate tasks in DAG for example.

Reply:
I'm not sure that I follow. Can you expand?

Reply:
Maybe to clarify: you will always have a nonce in the invocation payload. We just don't index on it, which is what this function is for.

Reply:
@expede, in older versions that nonce could be empty (i.e. an empty string), and, previously, the default was `""`. That should still be possible, right (I may have missed that), because it's required in the payload? Otherwise, you'll *probably* never be able to match on a task identifier for replay (unless you use the same nonce of course). 

Nonetheless, I think we want a default impl version that uses the Nonce (whatever it is, empty (`""`) or not) that's not just for DynJs structs. 
Otherwise, the trait should then make the Nonce optional if you don't want to use it at all.

Reply:
> Maybe to clarify: you will always have a nonce in the invocation payload. We just don't index on it, which is what this function is for.

Right, I get that. I'm saying we should default it for general use, b/c it's needed. Nonetheless, my previous question on Nonce default and empty string also is something to bring up.

Reply:
@zeeshanlakhani here's the latest version of the code (not yet pushed because stuff's moving)

```rust
// Don't get too hung up on the name
pub struct Task {
    pub sub: Option<Did>, // Is this optional? May as well make it so for now!
    pub nonce: Option<Nonce>,

    pub cmd: String,
    pub args: BTreeMap<String, Ipld>,
}
```

This is to be clearer than just returning a `Cid` — its saying "hey, there's the stuff that's going into a TaskId". We can decide how we want that info serialized to a CID.

Reply:
Yeah, This makes some sense, as `Task` needed to exist somewhere.